packages/@aws-cdk/aws-codepipeline-api/lib/action.ts

```diff
@@ -71,6 +71,25 @@ export interface IStage {
    * @param action the Action to add to this Stage
    */
   _attachAction(action: Action): void;
+
+  /**
+   * Generates a unique output artifact name for the given Action.
+   * This is an internal operation -
+   * you should never need to call it directly.
+   *
+   * @param action the Action to generate the output artifact name for
+   */
+  _generateOutputArtifactName(action: Action): string;
+
+  /**
+   * Finds an input artifact for the given Action
+   * among the existing output artifacts currently in the Pipeline.
+   * This is an internal operation -
+   * you should never need to call it directly.
```

Comment:
Provide some details on the algorithm

Reply:
Done (it's not an easy one to describe...).